crates/factor-variables/src/provider/env.rs

```diff
@@ -0,0 +1,215 @@
+use std::{
+    collections::HashMap,
+    env::VarError,
+    path::{Path, PathBuf},
+    sync::OnceLock,
+};
+
+use serde::Deserialize;
+use spin_expressions::{Key, Provider};
+use spin_factors::anyhow::{self, Context as _};
+use spin_world::async_trait;
+use tracing::{instrument, Level};
+
+use crate::MakeVariablesProvider;
+
+/// Creator of a environment variables provider.
+pub struct EnvVariables;
+
+impl MakeVariablesProvider for EnvVariables {
+    const RUNTIME_CONFIG_TYPE: &'static str = "static";
+
+    type RuntimeConfig = EnvVariablesConfig;
+    type Provider = EnvVariablesProvider;
+
+    fn make_provider(&self, runtime_config: Self::RuntimeConfig) -> anyhow::Result<Self::Provider> {
+        Ok(EnvVariablesProvider::new(
+            runtime_config.prefix,
+            |key| std::env::var(key),
+            runtime_config.dotenv_path,
+        ))
+    }
+}
+
+/// Configuration for the environment variables provider.
+#[derive(Debug, Default, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct EnvVariablesConfig {
+    /// A prefix to add to variable names when resolving from the environment.
+    ///
+    /// Unless empty, joined to the variable name with an underscore.
+    #[serde(default)]
+    pub prefix: Option<String>,
+    /// Optional path to a 'dotenv' file which will be merged into the environment.
+    #[serde(default)]
+    pub dotenv_path: Option<PathBuf>,
+}
+
+const DEFAULT_ENV_PREFIX: &str = "SPIN_VARIABLE";
+
+/// A config Provider that uses environment variables.
+pub struct EnvVariablesProvider {
+    prefix: Option<String>,
+    env_fetcher: Box<dyn Fn(&str) -> Result<String, VarError> + Send + Sync>,
+    dotenv_path: Option<PathBuf>,
+    dotenv_cache: OnceLock<HashMap<String, String>>,
+}
+
+impl EnvVariablesProvider {
+    /// Creates a new EnvProvider.
+    ///
+    /// * `prefix` - The string prefix to use to distinguish an environment variable that should be used.
+    /// If not set, the default prefix is used.
+    /// * `env_fetcher` - The function to use to fetch an environment variable.
+    /// * `dotenv_path` - The path to the .env file to load environment variables from. If not set,
+    /// no .env file is loaded.
+    pub fn new(
+        prefix: Option<impl Into<String>>,
+        env_fetcher: impl Fn(&str) -> Result<String, VarError> + Send + Sync + 'static,
```

Comment:
Do we have a concrete need for this? Seems a bit marginal vs implementing an entirely new provider...

Reply:
It makes testing not rely on the environment it's run in. If it complicates too many things in your opinion, we can go back to just assuming `std::env::variable`.

Reply:
Maybe a separate `with_env_fetcher`? :shrug: